larpandoracontent/LArMonitoring/SliceMonitoringTool.cc

```diff
@@ -0,0 +1,361 @@
+/**
+ *  @file   larpandoracontent/LArMonitoring/SliceMonitoringTool.cc
+ *
+ *  @brief  Implementation of the slice monitoring tool.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoracontent/LArMonitoring/SliceMonitoringTool.h"
+
+#include "larpandoracontent/LArObjects/LArCaloHit.h"
+
+#include "larpandoracontent/LArHelpers/LArGeometryHelper.h"
+#include "larpandoracontent/LArHelpers/LArMCParticleHelper.h"
+#include "larpandoracontent/LArHelpers/LArMvaHelper.h"
+#include "larpandoracontent/LArHelpers/LArPfoHelper.h"
+
+#include <algorithm>
+#include <iterator>
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+SliceMonitoringTool::SliceMonitoringTool() {}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+SliceMonitoringTool::~SliceMonitoringTool()
+{
+    PANDORA_MONITORING_API(SaveTree(this->GetPandora(), m_treename.c_str(), m_filename.c_str(), "UPDATE"));
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void SliceMonitoringTool::ProcessSlices(const pandora::Algorithm *const pAlgorithm, SlicingAlgorithm::SliceList &inputSliceList)
+{
+
+    const MCParticleList *pMCParticleList{nullptr};
+    PANDORA_THROW_RESULT_IF(
+        STATUS_CODE_SUCCESS, !=,
+        PandoraContentApi::GetCurrentList(*pAlgorithm, pMCParticleList));
+
+    const CaloHitList *pCompleteCaloHitList{nullptr};
+    PANDORA_THROW_RESULT_IF(
+        STATUS_CODE_SUCCESS, !=,
+        PandoraContentApi::GetList(*pAlgorithm, "FullHitList", pCompleteCaloHitList));
+
+    // Populate the complete calo hit list, based on every hit in every slice.
+    CaloHitList fullSliceCaloHitList{};
+    for (const auto &slice : inputSliceList)
+    {
+        for (const CaloHit *const pSliceCaloHit : slice.m_caloHitListU)
+            fullSliceCaloHitList.push_back(pSliceCaloHit);
+        for (const CaloHit *const pSliceCaloHit : slice.m_caloHitListV)
+            fullSliceCaloHitList.push_back(pSliceCaloHit);
+        for (const CaloHit *const pSliceCaloHit : slice.m_caloHitListW)
+            fullSliceCaloHitList.push_back(pSliceCaloHit);
+    }
+
+    const MCParticle *pTrueNeutrino{nullptr};
+    LArMCParticleHelper::MCContributionMap mcToTrueHitListMap;
+
+    // Since the slice hits, and the full, unfiltered hits may not match up, first
+    // perform a quick match between these two sets of hits.
+    std::map<const std::tuple<float, float, float>, const CaloHit*> caloHitListMatchMap;
+    auto getHitKey = [](const CaloHit* pCaloHit) -> std::tuple<float, float, float> {
+        const auto pos = pCaloHit->GetPositionVector();
+        return {pos.GetX(), pos.GetZ(), pCaloHit->GetHadronicEnergy()};
+    };
+    CaloHitList matchedCaloHitList;
```

Comment:
This is also a knock on impact of that strange hack.

The "full hit list" and the "actual hit list" are different underlying hits.
Since there isn't a way for Pandora to know they are the same (say if the hits had a built in `uuid` or something), I have to do a somewhat iffy map to go from "hit in full hit list" to "hit in actual hit list".

Its awkward, and ideally people wouldn't need to do that (assuming again, we want to keep a "full hit list" around).